spec.md

```diff
@@ -721,30 +721,6 @@ f(-1)           # returns 1 without printing
 ```
 
 
-It is a dynamic error for a function to call itself or another
-function value with the same declaration.
-
-```python
-def fib(x):
-  if x < 2:
-    return x
-  return fib(x-2) + fib(x-1)	# dynamic error: function fib called recursively
-
-fib(5)
-```
-
-This rule, combined with the invariant that all loops are iterations
-over finite sequences, implies that Starlark programs are not Turing-complete.
```

Comment:
@laurentlb: I have never liked the recursion check, and I wonder whether we can scrap it entirely when we (soon) have a cycle counter limit, which is really what we care about. At that point, we can let users express themselves with recursion when it is appropriate, and we can bound long-running programs regardless of whether they use recursive calls or just plain old loops.

Part of my objection is that it breaks function abstractions. Consider:
```
>>> def map(fn, list): return [fn(x) for x in list]
>>> def id(x): return x
>>> def f(x): map(id, [1,2,3])
>>> map(f, [4,5,6])
Traceback (most recent call last):
  <stdin>:1:4: in <expr>
  <stdin>:1:30: in map
  <stdin>:1:14: in f
  <stdin>:1:1: in map
Error: function map called recursively
```

The fact that f uses the map helper function means that it cannot itself be called, even indirectly, by the map function---even though there's no real recursion here. This is a spooky action at a distance.

I'm curious as to whether most users of go.starlark.net enable recursion, or leave it disabled (or don't care).  

I propose we implement the cycle limit sooner rather than later and eliminate the ban recursion and while loops. Then we advertise instead that "Starlark allows you to bound the number of computation steps of a thread".

Reply:
Yes, it breaks function abstractions. But in practice, this is rarely a problem. We encourage users to write list comprehensions instead of high order functions.

I agree we should implement a cycle limit, but it can be in addition to the current check.